File: swift/Sources/Aloud/Selection.swift
import AppKit
import ApplicationServices

/// Read the current selection from whatever app has focus.
///
/// Two paths, best first:
///  1. The Accessibility API — native text views expose their selection
///     directly, no clipboard involved.
///  2. A synthesised ⌘C round-trip — Electron apps (VS Code, Cursor, Slack,
///     Discord) don't answer the AX query, but they all answer Copy. The
///     clipboard is saved and restored around it.
enum SelectionCapture {
    enum CaptureError: Error {
        case notTrusted
    }

    static let copyTimeout: TimeInterval = 0.6

    static func capture(restoreClipboard: Bool) throws -> String {
        if let text = accessibilitySelection(), !text.isEmpty {
            return text.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return try clipboardSelection(restore: restoreClipboard)
    }

    // MARK: - AX path

    private static func accessibilitySelection() -> String? {
        let system = AXUIElementCreateSystemWide()
        var focusedRef: CFTypeRef?
        guard
            AXUIElementCopyAttributeValue(
                system, kAXFocusedUIElementAttribute as CFString, &focusedRef) == .success,
            let focusedRef,
            CFGetTypeID(focusedRef) == AXUIElementGetTypeID()
        else { return nil }
        let focused = focusedRef as! AXUIElement

        var selectedRef: CFTypeRef?
        guard
            AXUIElementCopyAttributeValue(
                focused, kAXSelectedTextAttribute as CFString, &selectedRef) == .success,
            let text = selectedRef as? String
        else { return nil }
        return text
    }

    // MARK: - Clipboard path

    private static func clipboardSelection(restore: Bool) throws -> String {
        guard AXIsProcessTrusted() else { throw CaptureError.notTrusted }

        let pasteboard = NSPasteboard.general
        let previous = pasteboard.string(forType: .string)
        let before = pasteboard.changeCount

        pressCopy()

        let deadline = Date().addingTimeInterval(copyTimeout)
        while Date() < deadline {
            if pasteboard.changeCount != before { break }
            Thread.sleep(forTimeInterval: 0.02)
        }
        guard pasteboard.changeCount != before else {
            // Nothing was selected, or the app ignored the keystroke.
            return ""
        }

        let text = pasteboard.string(forType: .string) ?? ""

        if restore, let previous {
            // Give the app a beat to finish writing before we overwrite it.
            Thread.sleep(forTimeInterval: 0.05)
            pasteboard.clearContents()
            pasteboard.setString(previous, forType: .string)
        }

        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func pressCopy() {
        let keyC: CGKeyCode = 8
        guard
            let down = CGEvent(keyboardEventSource: nil, virtualKey: keyC, keyDown: true),
            let up = CGEvent(keyboardEventSource: nil, virtualKey: keyC, keyDown: false)
        else { return }
        // Explicit flags: the user's hotkey modifiers may still be held.
        down.flags = .maskCommand
        up.flags = .maskCommand
        down.post(tap: .cghidEventTap)
        up.post(tap: .cghidEventTap)
    }
}

File: aloud/selection.py
"""Read the current selection from whatever app has focus.

macOS Services only reach apps built on native text views, which is why
Quick Actions silently do nothing in VS Code, Cursor, Slack, Discord and
every other Electron app. Synthesising a Copy keystroke works everywhere,
so that is what we do — then we put the user's clipboard back.
"""

from __future__ import annotations

import time

from AppKit import NSPasteboard, NSPasteboardTypeString
from Quartz import (
    CGEventCreateKeyboardEvent,
    CGEventPost,
    CGEventSetFlags,
    kCGHIDEventTap,
    kCGEventFlagMaskCommand,
)

KEY_C = 8
COPY_TIMEOUT = 0.6


def _press_copy():
    down = CGEventCreateKeyboardEvent(None, KEY_C, True)
    up = CGEventCreateKeyboardEvent(None, KEY_C, False)
    CGEventSetFlags(down, kCGEventFlagMaskCommand)
    CGEventSetFlags(up, kCGEventFlagMaskCommand)
    CGEventPost(kCGHIDEventTap, down)
    CGEventPost(kCGHIDEventTap, up)


def read_clipboard() -> str:
    pb = NSPasteboard.generalPasteboard()
    value = pb.stringForType_(NSPasteboardTypeString)
    return value or ""


def capture_selection(restore: bool = True) -> str:
    """Copy the focused selection and return it.

    Returns an empty string if nothing was selected — the clipboard change
    count tells us whether the app actually answered the Copy.
    """
    pb = NSPasteboard.generalPasteboard()
    previous = pb.stringForType_(NSPasteboardTypeString)
    before = pb.changeCount()

    _press_copy()

    deadline = time.time() + COPY_TIMEOUT
    while time.time() < deadline:
        if pb.changeCount() != before:
            break
        time.sleep(0.02)
    else:
        # Nothing was selected, or the app ignored the keystroke.
        return ""

    text = pb.stringForType_(NSPasteboardTypeString) or ""

    if restore and previous is not None:
        # Give the app a beat to finish writing before we overwrite it.
        time.sleep(0.05)
        pb.clearContents()
        pb.setString_forType_(previous, NSPasteboardTypeString)

    return text.strip()
